File: ios/SuperheroApp/Characters/Domain/UseCase/GetCharactersUseCase.swift
//
//  CharactersUseCase.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 12..
//  Copyright © 2018. W.UP. All rights reserved.
//

import Foundation
import Common

class GetCharactersRequest: CommonUseCaseRequest {
    var page: CommonPage

    init(page: CommonPage) {
        self.page = page
    }
}

class GetCharactersResponse: CommonUseCaseResponse {
    let characters: [CommonCharacter]

    init(characters: [CommonCharacter]) {
        self.characters = characters
    }
}

class GetCharactersUseCase: CommonUseCase {

    private let charactersDataSource: CommonCharactersDataSource

    init(charactersDataSource: CommonCharactersDataSource) {
        self.charactersDataSource = charactersDataSource
    }

    override func executeUseCase(request: CommonUseCaseRequest) {
        if let request = request as? GetCharactersRequest {

            self.charactersDataSource.loadCharacters(page: request.page, complete: { (characters: [CommonCharacter]) in
                let response = GetCharactersResponse(characters: characters)

                return self.success(response)
            }, fail: {
                return self.error()
            })
        }
    }
}

File: common/src/main/kotlin/digital/wup/superheroapp/common/characters/domain/usecase/GetCharactersUseCase.kt
package digital.wup.superheroapp.common.characters.domain.usecase

import digital.wup.superheroapp.common.UseCase
import digital.wup.superheroapp.common.UseCaseRequest
import digital.wup.superheroapp.common.UseCaseResponse
import digital.wup.superheroapp.common.characters.domain.model.Character
import digital.wup.superheroapp.common.characters.domain.model.Page
import digital.wup.superheroapp.common.datasource.CharactersDataSource

class GetCharactersRequest(val page: Page) : UseCaseRequest

class GetCharactersResponse(val characters: List<Character>): UseCaseResponse

class GetCharactersUseCase(private val charactersDataSource: CharactersDataSource): UseCase<GetCharactersRequest, GetCharactersResponse>() {

    override fun executeUseCase(request: GetCharactersRequest) {
        charactersDataSource.loadCharacters(request.page, { characters: List<Character> ->
            val response = GetCharactersResponse(characters)
            success(response)
        }, {
            error()
        })
    }
}